src/main/java/pl/allegro/tech/opel/OpelParsingResult.java

```diff
@@ -11,13 +11,23 @@
 public class OpelParsingResult {
     private final ParsingResult<ExpressionNode> parsingResult;
     private final String expression;
+    private final EvalContext embeddedEvalContext;
 
-    OpelParsingResult(String expression, ParsingResult<ExpressionNode> parsingResult) {
+    OpelParsingResult(String expression, ParsingResult<ExpressionNode> parsingResult, EvalContext embeddedEvalContext) {
         this.parsingResult = parsingResult;
         this.expression = expression;
+        this.embeddedEvalContext = embeddedEvalContext;
     }
 
     public CompletableFuture<?> eval(EvalContext context) {
+        return evalWithExactContext(EvalContext.Builder.mergeContexts(context, embeddedEvalContext));
```

Comment:
this is a private method, and context given in this method will not be merged with any other context.
Name of this method is not important to me, I was even thinking about evalPrivate ;)
I will change to evalWithFinalContext according to your proposal...